File: frontend/apps/uno/client/responses.ts
import { Card, DiscardedCard } from "./cards";
import { EventWithId } from "./events";

export type CreateResponse = {
  error: null;
  game_id: string;
  player_id: string;
};

export type DrawCardResponse = {
  error: null;
  events: EventWithId[];
};

export type GetEventsResponse = {
  error: null;
  events: EventWithId[];
};

export type GetStateResponse = {
  error: null;
  status: "abandoned" | "created" | "finished" | "started";
  players: {
    pid: string;
    avatar: {
      emoji: string;
      name: string;
    };
    cards: number;
    rounds_won: number;
    points: number;
  }[];
  player_pid: string;
  active_player_pid: string;
  admin_player_pid: string;
  next_event_id: number;
  cards: Card[];
  reversed: boolean;
  last_discard: DiscardedCard | null;
};

export type JoinResponse = {
  error: null;
  player_id: string;
};

export type LeaveResponse = {
  error: null;
};

export type ListResponse = {
  error: null;
  games: {
    id: string;
    status: "abandoned" | "created" | "finished" | "started";
    players: number;
  }[];
};

export type PlayCardResponse = {
  error: null;
  events: EventWithId[];
};

export type PlayWildCardResponse = {
  error: null;
  events: EventWithId[];
};

export type StartResponse = {
  error: null;
  events: EventWithId[];
};

File: backend/src/uno/responses.py
from typing import TypedDict

from .avatars import Avatar
from .cards import Card, DiscardedCard
from .constants import GameStatus
from .events import EventWithId


class CreateResponse(TypedDict):
    error: None
    game_id: str
    player_id: str


class DrawCardResponse(TypedDict):
    error: None
    events: list[EventWithId]


class GetEventsResponse(TypedDict):
    error: None
    events: list[EventWithId]


class Player(TypedDict):
    pid: str
    avatar: Avatar
    cards: int
    rounds_won: int
    points: int


class GetStateResponse(TypedDict):
    error: None
    status: GameStatus
    players: list[Player]
    player_pid: str
    active_player_pid: str
    admin_player_pid: str
    next_event_id: int
    cards: list[Card]
    reversed: bool
    last_discard: DiscardedCard | None


class JoinResponse(TypedDict):
    error: None
    player_id: str


class LeaveResponse(TypedDict):
    error: None


class Game(TypedDict):
    id: str
    status: GameStatus
    players: int


class ListResponse(TypedDict):
    error: None
    games: list[Game]


class PlayCardResponse(TypedDict):
    error: None
    events: list[EventWithId]


class PlayWildCardResponse(TypedDict):
    error: None
    events: list[EventWithId]


class StartResponse(TypedDict):
    error: None
    events: list[EventWithId]
